### PR TITLE
Feature 12 support for devices

### LINKED ISSUE #12: Support for device requests

### PR DESCRIPTION
**Pull request by villejuutila**
Resolves #12

### CODE REVIEW

File: src/test/kotlin/fi/metatavu/noheva/api/test/functional/builder/impl/ExhibitionDeviceTestBuilderResource.kt [hunk 1]
```diff
@@ -25,14 +25,12 @@ class ExhibitionDeviceTestBuilderResource(
      *
      * @param exhibitionId exhibition id
      * @param groupId group id
-     * @param modelId model id
      * @return created exhibition device
      */
-    fun create(exhibitionId: UUID, groupId: UUID, modelId: UUID): ExhibitionDevice {
+    fun create(exhibitionId: UUID, groupId: UUID): ExhibitionDevice {
```
Comment (anttileppa) [inline src/test/kotlin/fi/metatavu/noheva/api/test/functional/builder/impl/ExhibitionDeviceTestBuilderResource.kt:30]: Shouldn't we always include deviceId when creating exhibition devices?

File: src/main/kotlin/fi/metatavu/noheva/api/DevicesApiImpl.kt [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                ?: return createBadRequest("Device model ${device.deviceModelId} not found")
+        }
+
+        foundDevice.approvalStatus.let {
+            when (it) {
+                DeviceApprovalStatus.PENDING ->
+                    if (device.approvalStatus != it && device.approvalStatus != DeviceApprovalStatus.APPROVED) {
+                        return createBadRequest("Cannot change approval status")
+                    }
+                DeviceApprovalStatus.APPROVED ->
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/noheva/api/DevicesApiImpl.kt:133]: Why do we allow changing approval status from approved to pending?

**Review by anttileppa — CHANGES_REQUESTED** [review]

File: src/main/kotlin/fi/metatavu/noheva/devices/DeviceController.kt [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+     * @param userId user id
+     * @return updated device
+     */
+    fun updateDevice(
+        existingDevice: Device,
+        newDevice: fi.metatavu.noheva.api.spec.model.Device,
+        deviceModel: DeviceModel?,
+        userId: UUID
+    ): Device {
+        existingDevice.name = newDevice.name
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/noheva/devices/DeviceController.kt]: Resolved.

File: src/main/kotlin/fi/metatavu/noheva/devices/DeviceController.kt [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+     *
+     * @param device device
+     * @param key key
+     * @return updated device
+     */
+    fun storeDeviceKey(device: Device, key: PublicKey): Device {
+        device.deviceKey = key.encoded
+        device.approvalStatus = DeviceApprovalStatus.READY
+
+        return deviceDAO.update(device)
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/noheva/devices/DeviceController.kt]: Resolved.

File: src/main/kotlin/fi/metatavu/noheva/devices/DeviceController.kt [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+     * @return updated device
+     */
+    fun reInitiateDeviceApproval(device: Device, name: String?, description: String?, version: String): Device {
+        device.name = name
+        device.description = description
+        device.version = version
+        device.deviceKey = null
+        device.approvalStatus = DeviceApprovalStatus.PENDING
+
+        return deviceDAO.update(device)
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/noheva/devices/DeviceController.kt]: Resolved.

File: src/main/kotlin/fi/metatavu/noheva/persistence/dao/DeviceDAO.kt [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        return getEntityManager().createQuery(criteria).resultList
+    }
+
+    /**
+     * Updates Device
+     *
+     * @param device device
+     * @return updated device
+     */
+    fun update(device: Device): Device {
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/noheva/persistence/dao/DeviceDAO.kt]: Resolved.

File: src/main/kotlin/fi/metatavu/noheva/api/DevicesApiImpl.kt (hunk 2 above)
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/noheva/api/DevicesApiImpl.kt:133]: I thought that just in case someone would want to revert the approval

**Review by anttileppa — APPROVED** [review]

**Pull request merged**